weatherman/calculations.py

```diff
@@ -0,0 +1,146 @@
+from typing import Dict, Tuple, List
+from collections import defaultdict
+from weatherman.utils import extract_year, extract_year_month
+from weatherman.results import (YearlyExtremeWeatherResult, MonthlyAveragesWeatherResult,
+                                MonthlyExtremeBarChartResult)
+
+
+class WeatherCalculations:
+    """
+    Performs calculations on parsed weather data.
+    """
+    def __init__(self, weather_recordings_data: list[Dict[str, str]]):
+        self.weather_data = weather_recordings_data
+
+    def compute_yearly_extreme_weather(self, year: str) -> YearlyExtremeWeatherResult:
+        weatherdata_filtered_by_year = self.filter_weather_data_by_year(year)
+        max_temp_tuple: Tuple[float, str] = (float('-inf'), '')
+        min_temp_tuple: Tuple[float, str] = (float('inf'), '')
+        max_humid_tuple: Tuple[float, str] = (float('-inf'), '')
+        for weather_entry in weatherdata_filtered_by_year:
+            if weather_entry.get('Max TemperatureC'):
+
+                try:
+                    max_temp_value = float(weather_entry['Max TemperatureC'])
+                    if max_temp_value > max_temp_tuple[0]:
+
+                        max_temp_tuple = (max_temp_value, weather_entry['PKT'])
+                except ValueError:
+                    pass
+
+            if weather_entry.get('Min TemperatureC'):
+
+                try:
+                    min_temp_value = float(weather_entry['Min TemperatureC'])
+                    if min_temp_value < min_temp_tuple[0]:
+
+                        min_temp_tuple = (min_temp_value, weather_entry['PKT'])
+                except ValueError:
+                    pass
+
+            if weather_entry.get('Max Humidity'):
+
+                try:
+                    max_humid_value = float(weather_entry['Max Humidity'])
+                    if max_humid_value > max_humid_tuple[0]:
+
+                        max_humid_tuple = (max_humid_value, weather_entry['PKT'])
+
+                except ValueError:
+                    continue
+
+        return YearlyExtremeWeatherResult(
+            max_temp_tuple[0],
+            max_temp_tuple[1],
+            min_temp_tuple[0],
+            min_temp_tuple[1],
+            max_humid_tuple[0],
+            max_humid_tuple[1]
+        )
+
+    def compute_monthly_averages_weather(self, month: str) -> MonthlyAveragesWeatherResult:
+        weather_data_filtered_by_month = self.filter_weather_data_by_month(month)
+        sums, counts = defaultdict(float), defaultdict(int)
+        for weather_entry in weather_data_filtered_by_month:
+            try:
+                sums['highest'] += float(weather_entry.get('Max TemperatureC', 0))
+                counts['highest'] += 1
+            except ValueError:
```

Comment:
please find a generic way to raise your exceptions and catch them, properly log them, passing it won't help